### PR TITLE
feat: support trimmed, stripped and CGo binaries

### PR DESCRIPTION
**Pull request by tommyknows**
- [ ] Ready for review
- [x] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?

This commit updates our support for Go binaries built with the
`-trimpath` flag, stripped binaries (`-ldflags="-s -w"`), vendored
dependencies or with CGo enabled. This means that customers can finally
scan pretty much any Go binary (built with the Go compiler) and get
meaningful results 🎉

The new support is implemented by parsing the PCLN table and going
through the list of files that is in there. This list contains the paths
of the files which includes their module and package name. We extract
this to match them to the Go module version that we get from the
`.go.buildinfo` ELF section (same as what `go version -m <binary>`
does).

The actual implementation needs to check for a couple of edge-cases, as
trimmed binaries or vendored dependencies need to be treated slighly
differently than regular binaries.

The tests have been extended to support that as well, a vendored
binary has been added to the tests and the regression tests will always
build vendored binaries as well.

#### How should this be manually tested?

You can - don't think it's necessary - build the "latest" binaries locally (go to `test/fixtures/go-binaries/source`, call `build.sh latest`) and then run the tests.

#### What are the relevant tickets?

https://snyksec.atlassian.net/browse/MYC-146

### CODE REVIEW

**Comment by tommyknows** [comment]
I guess I need to mark it as a breaking change as support for non-module binaries has been removed? 
Or should we keep the old behaviour as a backup-mechanism if the binary wasn't built with modules? Or even keep it as the primary behaviour if the `strtab` exists and only fall back to the PCLN table stuff if the `strtab` does not exist?

**Comment by noaleibo1** [comment]
> In my opinion, not attempting to scan it is better

I agree. You can double check with Hadar, but I doubt she'll have a different opinion.

**Review by RotemS — COMMENTED** [review]
Nice work!! 🚀 💪 
Left some comments :)

File: lib/go-parser/go-binary.ts [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+import { eventLoopSpinner } from "event-loop-spinner";
+import * as path from "path";
+import * as varint from "varint";
+
+import { DEP_GRAPH_TYPE } from "./";
+import { GoModule } from "./go-module";
+import { LineTable } from "./pclntab";
+import { Elf, ElfProgram } from "./types";
+
+export class GoBinary {
```
[the remark this answers is not in the document]
Comment (tommyknows) [reply lib/go-parser/go-binary.ts:11]: `go-module` cotnains the `GoModule` class, not the `GoBinary` class 🤔

File: lib/go-parser/go-binary.ts [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+  // matchFilesToModules goes through all files, extracts the package name and
+  // adds it to the relevant module in the GoBinary.
+  public matchFilesToModules(files: string[]): void {
+    // goModCachePath is the path at which the modules are downloaded to. When
+    // building a Go binary, this is usually either $GOMODCACHE or
+    // $GOROOT/pkg/mod. Binaries built with `-trimpath` will have that module
+    // cache path trimmed away, meaning that the goModCachePath will always be
+    // empty.
+    let goModCachePath: string = "";
+    let moduleName = (mod: GoModule): string => {
```
[the remark this answers is not in the document]
Comment (tommyknows) [reply lib/go-parser/go-binary.ts:59]: It is because we need to use this function on all modules that we have. The module name is usually `my.modules.com/module@v1.0.0`, but in vendored binaries it is simply `my.modules.com/module`. 

This is for *all* modules in a binary, which is why we need to use a different function to get that name.

File: lib/go-parser/go-binary.ts [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+            // e.g. /app/vendor/github.com/my/repo/. As such, the "full name" of
+            // the module should also not contain the version string.
+            (m: GoModule): string => {
+              return m.name;
+            },
+          ];
+        }
+      }
+    }
+    return ["", fullName];
```
[the remark this answers is not in the document]
Comment (tommyknows) [reply lib/go-parser/go-binary.ts:163]: Similar as above - yes, because we don't do this on a specific module, but on all modules :)

File: lib/go-parser/go-binary.ts [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+      for (const module of this.modules) {
+        const modFullName = moduleName(module);
+        if (pkgFile.startsWith(modFullName)) {
+          // we try to split the filename into two sections, the "module"
+          // section and the "package and file" section.
+          // For example, the filename "github.com/my/pkg@v0.0.1/a/a.go" will be
+          // split into "github.com/my/pkg@v0.0.1/" and "a/a.go". We then get
+          // the package name from the package and file section, and add the
+          // normalized module name (without the version) in front. This will
+          // result in the package name "github.com/my/pkg/a".
```
[the remark this answers is not in the document]
Comment (tommyknows) [reply lib/go-parser/go-binary.ts:101]: Are there other comments that you think could be more succinct or even deleted completely? 

(I still want people to easily read through the code and understand *why* we're doing all this, even if they do not have any knowledge of how Go organises the files)

File: lib/go-parser/go-binary.ts [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+  const mod = readRawBuildInfo(binary);
+  if (!mod) {
+    throw Error("binary contains empty module info");
+  }
+
+  const [, mainModuleLine, ...versionsLines] = mod.split("\n");
+  const [, name] = mainModuleLine.split("\t");
+  const modules: GoModule[] = [];
+  versionsLines.forEach((versionLine) => {
+    if (!versionLine) {
```
[the remark this answers is not in the document]
Comment (tommyknows) [reply lib/go-parser/go-binary.ts:184]: I would expect that it's either empty or undefined, both of which would also be caught by the `split` / `if (!name || !ver)` expressions.

This code has been copied from the previous implementation, so I don't know why it's there, but I think we can remove it :)

File: test/unit/go-binaries.spec.ts [hunk 6]
```diff
@@ -0,0 +1,398 @@
+import * as depGraph from "@snyk/dep-graph";
+import * as elf from "elfy";
+import * as fs from "fs";
```
[the remark this answers is not in the document]
Comment (tommyknows) [reply test/unit/go-binaries.spec.ts:3]: Why is it read**d**irSync but read**F**ileSync 🙄🙄🙄🙄

[the remark this answers is not in the document]
Comment (RotemS) [reply test/unit/go-binaries.spec.ts:3]: I got annoyed by that too :D 

I originally copied just one and change `dir` to `file` but then realized it's wrong :D

**Comment by tommyknows** [comment]
So I checked the JSON output of a `snyk container test` on an image that contains a Go binary that we cannot scan. 

I cannot find anything in that JSON that refers to said binary, so we "skip" it. As such, behaviour to the end user is the same whether we scan it and return an empty scan result or not scan it at all. 

This means that this PR does not contain a breaking change 🎉

File: test/unit/go-binaries.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tommyknows) [reply test/unit/go-binaries.spec.ts:3]: There's a [reason](https://github.com/nodejs/node/issues/9664#issuecomment-261333191) for it, though I'd respectfully disagree :D

File: lib/go-parser/go-binary.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (RotemS) [reply lib/go-parser/go-binary.ts:101]: Don't want to block on comments :)

**Review by RotemS — DISMISSED** [review]

**Review by RotemS — APPROVED** [review]
